main.go

```diff
@@ -177,91 +192,225 @@ func main() {
-
-		if opts.EndpointRead != "" {
-			g.Add(func() error {
-				level.Info(logger).Log("msg", "start querying for metrics")
-				for {
-					select {
-					case <-t.C:
-						if err := read(ctx, endpointRead, opts.Labels, opts.Latency); err != nil {
-							queryResponses.WithLabelValues("error").Inc()
-							level.Error(logger).Log("msg", "failed to query", "err", err)
-						} else {
-							queryResponses.WithLabelValues("success").Inc()
-						}
-					case <-ctx.Done():
-						success, errors := successerrors(logger)
-						ratio := success / (success + errors)
-
-						if ratio < opts.SuccessThreshold {
-							return fmt.Errorf("failed with less than %2.f%% success ratio - actual %2.f%%", opts.SuccessThreshold*100, ratio*100)
-						}
-						return nil
-					}
-				}
-			}, func(err error) {
-				t.Stop()
-				cancel()
-			})
-		}
+		g.Add(func() error {
+			return runReader(ctx, opts, l, m)
+		}, func(_ error) {
+			cancel()
+		})
 	}
 
 	if err := g.Run(); err != nil {
-		stdlog.Fatal(err)
+		level.Error(l).Log("msg", "run group exited with error", "err", err)
+		os.Exit(1)
+	}
+
+	level.Info(l).Log("msg", "up completed its mission!")
+}
+
+func runReader(ctx context.Context, opts options, logger log.Logger, m metrics) error {
+	l := log.With(logger, "component", "reader")
+	level.Info(l).Log("msg", "starting the reader")
+
+	// Wait for at least one period before start reading metrics.
+	level.Info(l).Log("msg", "waiting for initial delay before querying for metrics")
+	select {
+	case <-ctx.Done():
+		return nil
+	case <-time.After(opts.InitialQueryDelay + opts.Period):
+	}
+
+	t := time.NewTicker(opts.Period)
+
+	level.Info(l).Log("msg", "start querying for metrics")
+
+	for {
+		select {
+		case <-t.C:
+			// Do not propagate parent context to prevent cancellation of in-flight request.
+			if err := read(context.Background(), opts.ReadEndpoint, opts.Labels, -1*opts.InitialQueryDelay, opts.Latency, m); err != nil {
+				m.queryResponses.WithLabelValues("error").Inc()
+				level.Error(l).Log("msg", "failed to query", "err", err)
+			} else {
+				m.queryResponses.WithLabelValues("success").Inc()
+			}
+		case <-ctx.Done():
+			t.Stop()
+			return reportResults(l, m.queryResponses, opts.SuccessThreshold)
+		}
+	}
+}
+
+func runWriter(ctx context.Context, opts options, logger log.Logger, m metrics) error {
+	l := log.With(logger, "component", "writer")
+	t := time.NewTicker(opts.Period)
+
+	level.Info(l).Log("msg", "starting the writer")
+
+	for {
+		select {
+		case <-t.C:
+			// Do not propagate parent context to prevent cancellation of in-flight request.
+			if err := write(context.Background(), opts.WriteEndpoint, opts.Token, generate(opts.Labels), l); err != nil {
+				m.remoteWriteRequests.WithLabelValues("error").Inc()
+				level.Error(l).Log("msg", "failed to make request", "err", err)
+			} else {
+				m.remoteWriteRequests.WithLabelValues("success").Inc()
+			}
+		case <-ctx.Done():
+			t.Stop()
```

Comment:
In this approach the process won’t wait for requests to finish

Reply:
So since this is in a select in a single goroutine, I assumed when it execution comes in this branch there aren't any requests in-flight. Although the ticker doesn't close its channel, it'd exit after this instruction, so it should be fine.

Reply:
Hmm yes you’re right. The real problem is that we may wait infinitely for the requests to finish. The easy way around this would be to use a context with timeout on the ctx of the request where the timeout is the period. So we will wait at most 1 extra period for a request.

Reply:
this was what I did exactly. Then I thought all the actors will be stopped, however, when  I check it again this is not the case.

I'll fix it.